yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala

```diff
@@ -121,13 +121,20 @@ private[spark] abstract class YarnSchedulerBackend(
     }
   }
 
+  private[cluster] def prepareRequestExecutors(requestedTotal: Int): RequestExecutors = {
+    val nodeBlacklist: Set[String] = scheduler.nodeBlacklist()
+    val filteredHostToLocalTaskCount =
+      hostToLocalTaskCount.filter { case (k, v) => !nodeBlacklist.contains(k) }
```

Comment:
actually, this doesn't have anything to do with yarn really -- its just that only on yarn does dynamic allocation use locality preferences (`hostToLocalTaskCount`).  We want locality preferences to ignore preferences for blacklisted nodes.

this could alternatively be done in `YarnAllocator`, when it receives this msg: https://github.com/apache/spark/blob/master/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala#L227

It could also be done in `CoarseGrainedSchedulerBackend` before it delegates to specific impls (like YarnSchedulerBackend), eg https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala#L507

the benefit of that is it would be easier to make its way into standalone & mesos.  But it also calls `doRequestExecutors` multiple places, so you still need to make sure you do this filter in all those spots, to ensure you're always using the most up-to-date blacklist.